presto-main/src/main/java/com/facebook/presto/operator/project/MergingPageOutput.java

```diff
@@ -0,0 +1,182 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.project;
+
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.PageBuilder;
+import com.facebook.presto.spi.block.ArrayBlockBuilder;
+import com.facebook.presto.spi.type.Type;
+import com.google.common.collect.ImmutableList;
+import org.openjdk.jol.info.ClassLayout;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.NotThreadSafe;
+
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Queue;
+
+import static com.facebook.presto.spi.block.PageBuilderStatus.DEFAULT_MAX_PAGE_SIZE_IN_BYTES;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * This class is intended to be used right after the PageProcessor to ensure
+ * that the size of the pages returned by FilterAndProject and ScanFilterAndProject
+ * is big enough so it does not introduce considerable synchronization overhead.
```

Comment:
Is it really only about synchronization?

Just having very small pages offsets any benefits of vectorization since the page overhead of processing single row becomes significant. Presto isn't really written to process `Pages` fast (vide `Driver#processInternal`), but rather process rows within pages fast.